# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data 

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [18]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced
from sklearn.metrics import balanced_accuracy_score

import warnings
warnings.filterwarnings('ignore')

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv(
    Path("Resources/lending_data.csv"))

# Review the DataFrame
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
labels = lending_data_df['loan_status']

# Separate the X variable, the features
features = lending_data_df[['loan_size','interest_rate','borrower_income','debt_to_income',
                            'num_of_accounts', 'derogatory_marks', 'total_debt']]

In [4]:
# Review the y variable Series
labels.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
features.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
labels.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
training_features, testing_features, training_labels, testing_labels = train_test_split(features,labels)

training_features

loan_size  interest_rate  borrower_income  debt_to_income  \
60734     9200.0          7.046            46900        0.360341   
2962      9800.0          7.282            49100        0.389002   
50194    10000.0          7.361            49900        0.398798   
45126     9900.0          7.346            49700        0.396378   
41695    10700.0          7.654            52600        0.429658   
...          ...            ...              ...             ...   
76419    17400.0         10.507            79500        0.622642   
34095     8300.0          6.662            43300        0.307159   
70300     8800.0          6.855            45100        0.334812   
60152     9300.0          7.086            47300        0.365751   
475       9600.0          7.195            48300        0.378882   

       num_of_accounts  derogatory_marks  total_debt  
60734                3                 0       16900  
2962                 4                 0       19100  
50194                4                 0       19900  
45126                4                 0       19700  
41695                5                 1       22600  
...                ...               ...         ...  
76419               11                 2       49500  
34095                2                 0       13300  
70300                3                 0       15100  
60152                3                 0       17300  
475                  4                 0       18300  

[58152 rows x 7 columns]

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
lr = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = lr.fit(training_features, training_labels)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
x_test = lr_model.predict(testing_features)

results_df = pd.DataFrame({"Prediction":x_test, "Actual":testing_labels})

results_df


Prediction  Actual
12860           0       0
55637           0       0
11486           0       0
33897           0       0
77136           1       1
...           ...     ...
48791           0       0
69504           0       0
25847           0       0
54047           0       0
30010           0       0

[19384 rows x 2 columns]

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Print the balanced_accuracy score of the model
accuracy_score(testing_labels, x_test)

0.9916941807676434

In [11]:
# Generate a confusion matrix for the model
confusion_matrix(testing_labels, x_test)

array([[18677,    96],
       [   65,   546]], dtype=int64)

In [12]:
# Print the classification report for the model
print(classification_report_imbalanced(testing_labels, x_test))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.89      1.00      0.94      0.90     18773
          1       0.85      0.89      0.99      0.87      0.94      0.88       611

avg / total       0.99      0.99      0.90      0.99      0.94      0.90     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** For this test data: Accuracy looks extremely look- approximately 99% of the transactiosn in the test data were accurately categorized by the model. However we should note that when using the value_counts feature only 3% of loans were unhealthy loans so even if the model predicted every loan as health it would still be a 97% accuracy rate

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [13]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_oversampler = RandomOverSampler(random_state = 1)


# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = random_oversampler.fit_resample(training_features, training_labels)

In [14]:
# Count the distinct values of the resampled labels data
y_resampled.value_counts()

0    56263
1    56263
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [16]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
model.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
y_pred = model.predict(testing_features)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [19]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_score(testing_labels, y_pred)

0.9938502060494827

In [21]:
# Generate a confusion matrix for the model
print(confusion_matrix(testing_labels, y_pred))

[[18665   108]
 [    4   607]]


In [22]:
# Print the classification report for the model
print(classification_report_imbalanced(testing_labels, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.99      1.00      0.99      0.99     18773
          1       0.85      0.99      0.99      0.92      0.99      0.99       611

avg / total       1.00      0.99      0.99      0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** this model correctly predicts the 0 100% of the time and correctly preditcs the 1 85% of the time